src/proto.rs

```diff
@@ -675,6 +734,45 @@ impl TryFrom<StructureTag> for LdapFilter {
 
                 Ok(LdapFilter::Equality(a, v))
             }
+            4 => {
+                let mut inner = value.expect_constructed().ok_or(())?;
+                inner.reverse();
+
+                let ty = inner
+                    .pop()
+                    .and_then(|t| t.match_class(TagClass::Universal))
+                    .and_then(|t| t.match_id(Types::OctetString as u64))
+                    .and_then(|t| t.expect_primitive())
+                    .and_then(|bv| String::from_utf8(bv).ok())
+                    .ok_or(())?;
+
+                let f = inner
+                    .pop()
+                    .and_then(|t| t.match_class(TagClass::Universal))
+                    .and_then(|t| t.match_id(Types::Sequence as u64))
+                    .and_then(|t| t.expect_constructed())
+                    .and_then(|bv| {
+                        let mut res: Vec<LdapSubstringFilter> = vec![];
+                        for StructureTag { class, id, payload } in bv.iter() {
+                            res.push(match (id, payload) {
```

Comment:
I don't think this quite correct. A substring filter can have 0 or 1 Initial, and 0 or 1 Final, and 0 to infinite Any. So I think that the content of "Substring(String, Vec<LdapSubstringFilter>)," should change to reflect this, perhaps it should have a structure for LdapSubstringFilter, which has:

```
struct LdapSubstringFilter {
    initial: Option<String>,
    any: Vec<String>,
    final: Option<String>,
}
```

And this parser should reject any substring filter where initial/final occur more than once.

Reply:
Done.